Improve pipeline speed and fix root certificate expiration issue

rija opened the pull request:
# Fix fallout of Let's Encrypt root an intermediate certificate expiration

Let's Encrypt DST Root CA X3 certificate expired on 30 September 2021.
Normally TLS libraries should have switched automatically to the existing valid replacement.
However the version of OpenSSL (1.1.0l) in Debian 9 (Stretch) had a bug preventing switching to the new certificate.
This caused composer to fail to connect.
The fix is to migrate our PHP containers' base image from Debian 9 (Stretch) to Debian 10 (Buster).
The later has version 1.1.1d of OpenSSL which doesn't present that buggy behaviour.
However moving to major new version of base image is not trivial a few stuff broke and needed to be fixed.
ALL breakage were fixed except for ``css_check`` which I've commented out for now until we figure why it doesn't work.
Client libraries for PostgreSQL 9.6 are no longer available in Debian 10, so I pulled version 11 of PostgreSQL client libraries. 
There doesn't appear to be problem with connecting to PostgreSQL 9.6 server using that version.

The composer error can be replicated with:
```
> docker-compose exec application curl https://asset-packagist.org
curl: (60) SSL certificate problem: certificate has expired
```

Another problem is Ansible playbook. The following error was thrown when running a playbook that use the ``yum`` module:
```
 Request failed: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:618)>
```

The temporary fix would have been  to disable certs validation as an option to the Ansible ``yum`` module call, but 
that would reduce the security of our systems.
The proper fix is to upgrade Centos used as the OS on the EC2 instances from Centos 7 to Centos 8.4.
That required changes:

* In the ``docker-install`` role we need to remove the "enable Docker-CE repo" step otherwise error happens
* In the ``postgres-preinstall`` role we need to install the Centos 8 PostgreSQL repo
* In the ``ansible-role-postgresql`` role configuration in ``playbook.yml``, we need to use PostgreSQL 11 and enable sudo with ``become``

# Improve caching in GitLab jobs and Docker build to speed up GitLab jobs

We have caching already for our custom images. Several other techniques are now applied (or restored):

## Local cache of base image from Docker Hub

the base image used in our containers are pulled from Docker Hub, but because it's GitLab job is isolated and create its own instance of ``docker-dind``,
the base are never available locally when a ``docker build`` command is triggered, requring pulling them for each job.

What we do is create a new preliminary GitLab stage (``.pre``), where we have a job to pull once all the base image we use in the project. We then save them as a TAR file archive. We then use ``GitLab artifacts`` functionality to make those files available to all subsequents stages.
In the jobs that need to build container images, we front the jobs' steps with a few line to load the TAR archive as local docker images so the build process doesn't need to pull them remotely.

## Authenticated login to Docker Hub

Until now, all the pull to Docker Hub are anonymous but they have rate limits for the number of pulls per period of time and these rates are different for anonymous users, for logged in free user and for paying users.
By logging with our Docker Hub account (which has to be set in GitLab variables) we increase our pull capacity.

## Caching of composer libraries

Since we have locked version with composer.lock, the vendor library can be cached between jobs.
To do so, we use the ``GitLab cache`` functionality that make a list of paths (in our case Composer files and directories) available across jobs, stages and pipelines of the same project.

We had that configuration before, but it disappeared, so we restore it.

## Fixed precise versions of base image

Until now, we tend to use ``latest`` or ``x.y`` (major version) as image tag when specifying base image for our Dockerfile.
The problem is that those container image can be updated whenever a minor version is released causing our cached image to be invalidated and trigger their pull and rebuild of our custom images.
Additionally, we don't have certainty on which version this loose tags base images are at, as the upgrade is not audited on our side.

Instead, we use precise tag for the base image (x.y.z) so to remove any chance for the base image to change.
We will manage upgrade of our infrastructure ourselves with our own auditing.

## Comment out docker pull and build instructions related to FUW

the container services associate with File Upload Wizard are not deployed on production environment, so there's no need
to pull and build their images.

## Fix tag for custom images

The production images built in the build job are tagged with the environment they are for.
Unfortunately, when they were pulled before the build the wrong tag  (``latest`` instead of ``stagging`` or ``live``) was used, causing the ``docker build`` to think there is no image so it build  entirely new production images again.
An associatated bug was the absence of the environment specific tag in ``docker-compose-build.yml`` where we define what image to use as cache.

## Move transcient docker instructions at the end of the Dockefile

In the ``Production-Web-Dockerfile``,  the block ``RUN apk ...`` was triggered every time causing extra compile time becuase previous layer is constantly invalidated by definition (the block creating site config which never persist). 
By moving the ``RUN apk ...`` before the Site config block we enable the compilation stage to be cached.

# Tests in CI

For the last few weeks we've made a lot of change to the infrastructure without running the test suites in CI, when I reactivated the test job, it failed mostly because of those changes, so the test job had to be adapted. The main change is that we now use the ``up.sh`` project setup script on CI test job as well.
I've also added an <exclude> block to functional tests in ``protected/tests/phpunit.xml`` to exclude flaky tests and those related to File Upload Wizard.

The other change is that Composer installed binaries like ``phpunit``, ``beat``,  ``phpcov`` need to be fully referenced because on CI we don't have the ``bin/`` symlink in project directory.

Finally the main gigadb test suite CI job is now re-instated.

# Bug fix

Fix the ``pg_dump`` command in ``convert_production_db_to_latest_ver.sh`` to use the version used by @pli888

TODO:

* [x] looking at build logs, I see that a lot of times is spent building the php base image, these changes will this happen only once per job, as the container is then available locally when building all our container image
* [x] by logging with Docker Hub account credentials, the process is no longer anonymous and it's highly possible that Docker shapes the bandwidth to their servers differently wether we are anonymous or logged in (not least because they have set up rate limits for docker pulls)
* [x] Our pipelines already use artifacts to pass data between job, we will pass docker base images the same way
* [x] The credentials for Docker Hub account (username and access token) need to be stored in GitLab varfiables set for the "All (default)" environment option
* [x] Fix root certificate issue that stems from [Let's Encrypt root certificate expiration](https://community.letsencrypt.org/t/help-thread-for-dst-root-ca-x3-expiration-september-2021/149190/51)

rija committed:
Speed up build stage by pulling php base image beforehand

Once pulled, it stays available for the duration of the step.
Also by login to Docker hub, you probably gets better bandwith than anonymous users.

Updated the CICD docs accordingly

rija committed:
Update the docs to say DOCKER_HUB_* are set for the All environment

as they are needed on CI environment as well as staging and live
and the value is the same for all those environments.

rija committed:
Add FUW base image to the before_script logged in pull

rija committed:
Pass base image between GitLab jobs using artifacts

we have a preparatory stage where we pull all base images and then save them as files
which are registered as GitLab job artifacts.

Then in all jobs that need to pull base images from Docker hub,
we load the image back from files so they become locally available to our containers without
requiring further pulls from Docker Hub.

rija committed:
remove an unnecessary dependency connection

rija committed:
add debug stage

rija committed:
Fix root certificate expiration issues in php-fpm debian base image

composer fails with error:
curl error 60 while downloading https://asset-packagist.org/packages.json: SSL certificate problem: certificate has expired

this due to a root certificate expiring an a buggy version of openssl.

The fix is to upgrade the version of debian used from stretch to buster.

rija committed:
Disable flaky tests and use up.sh on CI

rija committed:
Re-enable build for test job

rija committed:
Use up.sh on CI as well

the bin symlink don't exist on CI, so use full path to binaries for composer binaries.

also ensure composer.lock is accessible to the application container service

rija committed:
Make sure node base image and alpine are pre-loaded from Docker Hub

rija committed:
Add in cache base image for node and composer files

rija committed:
Update how csslint and jshint are built and run

to adapt to using Debian 10 buster as base image.

downloading with curl fails with certificate issues so use wget when building
Dockerfile.

rija committed:
Fix how css check an js check are called in test suite

rija committed:
Correct path to csslint

rija committed:
Disable ccs check temporarily as it's failing on CI

and I don't know yet why

rija committed:
Harmonise node base image

so the same version is used across the board

rija committed:
Fix path to a composer binary in coverage runner

rija committed:
Disable cert validation temporary for Ansible yum module

Fixes for now the
 Request failed: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:618)>
 error
 until we got time to do the proper fix

rija committed:
Update EC2 instance's OS from Centos 7 to Centos 8.4

in order to fix the `` certificate verify failed`` error thrown by Ansible
due expiration of LE's DST Root CA X3  root/intermediate certificate on 30 sept 2021.
This is done by changing the AMI id in the ``aws-instance.tf`` module.

The upgrade broke a couple of roles that needed fixed:
* In the ``docker-install`` role we need to remove the "enable Docker-CE repo" step otherwise error happens
* In the ``postgres-preinstall`` role we need to install the Centos 8 PostgreSQL repo
* In the ``ansible-role-postgresql`` role configuration in ``playbook.yml``, we need to use PostgreSQL 11 and enable sudo with ``become``

rija committed:
Use precise version number in base image instead of major version

so that they don't get rebuilt for each minor version

rija committed:
Use precise version number in base image instead of major version

so that they don't get rebuilt for each minor version

we've only done production container in previous commit. Now doing it for dev container

Also add base image caching for php:7.2.34-cli-buster I forgot to take into account

rija committed:
Fix missing base image

because I forgot to add the tar file in artifacts

rija committed:
Fix postgresql client version to match what's in Debian 10 buster

rija committed:
Fix docker image for gigadb worker pulling the wrong base image

rija committed:
Workers should use the PHP cli base image

rija committed:
Disable pulling and building images for FUW on production

because we are not deploying FUW for now

rija committed:
disable cache and artifacts for GitLab operation buttons

not needed (these are remote docker-compose operations) and take time to process making the job longer than necessary

rija committed:
Fix pg_dump arguments to use parameters in more consistent ways

noticed by @pli888, using the command he has used successfully

rija committed:
Comment more FUw specific steps in build job

rija committed:
Fix tag causing out of date production image to be pulled

rija committed:
Fix another tag causing out of date production image to be pulled

rija committed:
Tag newly pulled custom images so they server as cache to local build

rija committed:
Upload artifacts only on failure

it take times to upload artifacts at the end of a job.
They are only useful to investigate failures

rija committed:
Preserve intermediary layers of production image

rija committed:
Use recent version of docker-compose

as the previous one doesn't have the build --no-rm option

rija committed:
try install docker-compose with no version specified

rija committed:
show docker images after tag reassignment

rija committed:
Fix tag in build override of docker-compose to use environment

rija committed:
Reorder instructions in Web production dockerfile to optimise caching

the block that always change (and therefore invalidate the docker cache)
needs to be at the end so previous layer don't need to be rebuilt

rija committed:
Remove a stray RUN command

rija committed:
Re-enable test job for main app

kencho51 commented on up.sh:
```diff
@@ -80,6 +83,3 @@ docker-compose up -d fuw-worker gigadb-worker
 docker-compose run --rm test ./protected/yiic generatefiletypes
 docker-compose run --rm test ./protected/yiic generatefileformats
 
```
Should we include `docker-compose up -d portainer` as well?

rija replied:
@kencho51 CI environemtn is too transcient (it disappears after the test job has run) so portainer is not helping here

rija replied:
downloading artifacts is the procedure for debuging CI jobs. we need to make sure that the artifacts contains useful information to help with debugging

kencho51 commented on ops/infrastructure/roles/postgres-preinstall/tasks/main.yml:
```diff
@@ -1,6 +1,5 @@
-    name: https://download.postgresql.org/pub/repos/yum/reporpms/EL-7-x86_64/pgdg-redhat-repo-latest.noarch.rpm
-    
+    name: https://download.postgresql.org/pub/repos/yum/reporpms/EL-8-x86_64/pgdg-redhat-repo-latest.noarch.rpm
     state: present
```
I got this error when I tried to `ansible-playbook -i ../../inventories -i name_gigadb_server_staging_ken playbook.yml`
```
TASK [../../roles/postgres-preinstall : install PostgreSQL 11 repository] ******************************************************************************
fatal: [18.163.173.216]: FAILED! => {"changed": false, "msg": "Failed to validate GPG signature for pgdg-redhat-repo-42.0-20.noarch"}

PLAY RECAP *********************************************************************************************************************************************
18.163.173.216             : ok=60   changed=28   unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   
```
But when I appended `disable_gpg_check: yes` to this main.yml, then the ansible script could pass.
Reference could be found at [here](https://docs.ansible.com/ansible-core/devel/porting_guides/porting_guide_base_2.10.html).

pli888 replied:
I have this problem now after fixing the missing Gitlab variables problem.

rija replied:
@pli888 @kencho51 restarted the playbook and the error disappeared

pli888 replied:
I did a fresh install of the `ansible-role-postgresql` role by deleting it from the `~/.ansible/roles` directory then executing `$ ansible-galaxy install -r requirements.yml` from the `<path to>/gigadb-website/ops/infrastructure` directory. Running `$ ansible-playbook -i ../../inventories -i name_gigadb_server_staging_Peter playbook.yml` then installs PostgreSQL 11 in the EC2 dockerhost server.

kencho51 replied:
Hi @pli888, I have tried your steps to deploy a live server, but it still got the same error:
```
TASK [../../roles/postgres-preinstall : install PostgreSQL 11 repository] ******************************************************************************
fatal: [16.162.165.233]: FAILED! => {"changed": false, "msg": "Failed to validate GPG signature for pgdg-redhat-repo-42.0-20.noarch"}

PLAY RECAP *********************************************************************************************************************************************
16.162.165.233             : ok=60   changed=28   unreachable=0    failed=1    skipped=1    rescued=0    ignored=0   
```

rija committed:
Ensure .secrets is saved in artifacts after test job

rija committed:
Show list of running containers at the end of up.sh run

to help with showing context when debugging

rija committed:
Update up.sh to show all running containers and limit portainer to mac

portainer makes no sense running in the transient CI test run
and we need to know all running containers, not just the one managed by docker-compose

pli888 commented:
Line 72 in `aws-instance.tf` should be:
```
System = "t3_micro-centos8",
```

pli888 reviewed (approved):
I confirm that the code in this `faster-builds` branch allows a GigaDB website to be deployed on a staging server.

rija committed:
Change tag for main ec2 instance to show centos8 use

rija commented:
> t3_micro-centos8

Thanks @pli888. I've made that change

rija committed:
Add explicitly gpg key for the postgresql 11 repository

to avoid having the GPG signature verification error
